KAFKA-18402: Optimise of comparison on org.apache.kafka.common.Uuid

bmscomp opened the pull request:
Replace `if else ` conditions with `Long.compare` for reducing verbosity 
Add entry in .gitignore file for `core/data2` directory 
Fix some typos in javadoc 

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

divijvaidya commented on clients/src/main/java/org/apache/kafka/common/Uuid.java:
```diff
@@ -143,12 +143,6 @@ public int compareTo(Uuid other) {
             return 1;
```
Can we use `Long.compare()` for mostSignificantBits as well?

bmscomp replied:
@divijvaidya  I could not find a prettier way to use `Long.compare` for `mostSignificantBits` , do you have a trick for it ?

divijvaidya replied:
Something like this:
```
@Override
public int compareTo(Uuid other) {
    int res = Long.compare(mostSignificantBits, other.mostSignificantBits);
    return (res != 0) ? res : Long.compare(leastSignificantBits, other.leastSignificantBits);
}
```

bmscomp replied:
Thanks so much @divijvaidya,  looks great, I already updated the pull request

bmscomp replied:
@divijvaidya  Still there is another way to implement the `compareTo` method, and by the way `compare`  can be a pretty name also for the method, and it will keep the same style of the compare method name in `Long` 

 
       if (mostSignificantBits == other.mostSignificantBits)
            return Long.compare(leastSignificantBits, other.leastSignificantBits);
        return Long.compare(mostSignificantBits, other.mostSignificantBits);

divijvaidya commented on .gitignore:
```diff
@@ -35,6 +35,7 @@ Vagrantfile.local
 config/server-*
 config/zookeeper-*
 core/data/*
+core/data2/*
```
perhaps add a note that this is created by a test [1], other wise other reviewers may wonder (like me) as to where is this directory coming from

https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala#L2673

bmscomp replied:
@divijvaidya  Yes it comes from this test

ijuma replied:
Instead of doing this, we may want to fix that test to use tmp instead.

divijvaidya reviewed (approved):
Looks good to me!

bmscomp committed:
add missing closing parenthesis to java doc

bmscomp committed:
use Long.compare for mostSignificantBits

bmscomp committed:
add .gitignore entry for core/data2 directory

bmscomp committed:
complify conditions

bmscomp committed:
add comma and </p> to java doc

bmscomp committed:
fix some typos in javadoc

divijvaidya commented:
@bmscomp a lot of tests are failing right. Can you please rebase with latest trunk and we can actively try to merge this in next few days.

bmscomp commented:
@divijvaidya  Sorry for this delay, I just rebased the branch right now

divijvaidya commented:
I am apprehensive about making this change. Even JDK's implementation of UUID.compareTo() uses existing code:
```
@Override
    public int compareTo(UUID val) {
        // The ordering is intentionally set up so that the UUIDs
        // can simply be numerically compared as two numbers
        return (this.mostSigBits < val.mostSigBits ? -1 :
                (this.mostSigBits > val.mostSigBits ? 1 :
                 (this.leastSigBits < val.leastSigBits ? -1 :
                  (this.leastSigBits > val.leastSigBits ? 1 :
                   0))));
    }
```

My main concern is that the comparison is dependent on treating unsigned vs. signed values. for example, why not use Long.compareUnsigned instead of using Long.compare.

I would say, it's probably safer to stick to what JDK already does.

bmscomp committed:
use Long.compareUnsigned method instead of Long.compare

bmscomp commented:
@divijvaidya I think it's better to stick to the same implementation as JDK

bmscomp commented:
@divijvaidya  Do you think that this pull request worths a follow ? Please can you provide any feedback about it  thanks

bmscomp reviewed (commented):
@divijvaidya  review updated and the pull request already approved do you think it worth a merge ?

divijvaidya commented:
Sure, let's try to merge this in.

1. Can you please copy the openJDK's implementation of comparison?
2. Add a comment specifying that we are using implementation similar to openJDK.
3. Resolve merge conflicts.

bmscomp committed:
use the same implementation as official open jdk one

bmscomp commented:
@divijvaidya  This pull request is updated I used the  openJDK's implementation of comparison for the compareTo method of Uuid

ijuma commented on clients/src/main/java/org/apache/kafka/common/Uuid.java:
```diff
@@ -151,19 +151,22 @@ private byte[] getBytesFromUuid() {
+    /**
+     * Compares this Uuid with another given Uuid.
+     * <p>
+     * This implementation is taken from the compareTo method of the official java.util.UUID implementation.
```
We can't take the implementation from OpenJDK, that code has a difference license.

bmscomp replied:
@ijuma  We just called it openJDK implementation it's not specific to open JDK since  it's a recent one it's just about a way of comparison mean the compareTo, the same question can be asked when it comes to implement equals and hashCode for example every developer is getting inspired by the openJDK reference implementation 

@divijvaidya  What are your thoughts ? I think we have a nice topic right here on reference implementations and also Licences 

May be we have to add an entry on LICENCES file as we are using other libraries are referencing them 

I agree it's a very good question @ijuma

ijuma replied:
> May be we have to add an entry on LICENCES file as we are using other libraries are referencing them

That's only allowed when the license is compatible - OpenJDK's license is not.

> the same question can be asked when it comes to implement equals and hashCode for example every developer is getting inspired by the openJDK reference implementation

Please check https://github.com/apache/kafka/blob/trunk/CONTRIBUTING.md. It states clearly that you can only contribute your own work.

bmscomp replied:
@ijuma  Thanks so much for the review, please give me some time and I will come with my own optimisation, may be I need to stick on the early one I provided, I will update come back soon with some, I think I need some learnings to do this pull request and make it in a better shape

ijuma commented:
> Can you please copy the openJDK's implementation of comparison?

Please do not ever do this. The OpenJDK code has a different license.

bmscomp committed:
remove inspired comment from openJDK

bmscomp commented:
@ijuma @divijvaidya  I need some time to come back with my own code, I need some time for some learnings, as the topic is already amazing for me and it touches the contributions aspect and the license one

bmscomp committed:
fix some typos in java doc

bmscomp committed:
less verbose to list method

bmscomp committed:
less verbose toArray method

remove bad formatting

restore old formatting

bmscomp commented:
@ijuma  I already updated the pull request and changed the title, I came with a new implementation that is far from the openJDJK one, I also made  toList less verbose and also toArray 

Please let me know if any comment

bmscomp committed:
review getBytesFromUuid method

divijvaidya commented:
> > Can you please copy the openJDK's implementation of comparison?
> 
> Please do not ever do this. The OpenJDK code has a different license.

My bad for not checking the license. Thank you for catching this.

bmscomp commented:
@divijvaidya  @ijuma  Still think this is a better implementation, I think what do you think about it ? Or should I close this
